libcxx/test/std/experimental/reflection/p2996-ex-universal-formatter.pass.cpp

```diff
@@ -0,0 +1,87 @@
+//===----------------------------------------------------------------------===//
+//
+// Copyright 2024 Bloomberg Finance L.P.
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03 || c++11 || c++14 || c++17 || c++20
+// ADDITIONAL_COMPILE_FLAGS: -freflection
+// ADDITIONAL_COMPILE_FLAGS: -Wno-inconsistent-missing-override
+
+// <experimental/reflection>
+//
+// [reflection]
+
+#include <experimental/meta>
+#include <format>
+#include <print>
+
+namespace __impl {
+  template<auto... vals>
+  struct replicator_type {
+    template<typename F>
+      constexpr void operator>>(F body) const {
+        (body.template operator()<vals>(), ...);
+      }
+  };
+
+  template<auto... vals>
+  replicator_type<vals...> replicator = {};
+}
+
+template<typename R>
+consteval auto expand(R range) {
+  std::vector<std::meta::info> args;
+  for (auto r : range) {
+    args.push_back(reflect_value(r));
+  }
+  return substitute(^__impl::replicator, args);
+}
+
+struct universal_formatter {
+  constexpr auto parse(auto& ctx) { return ctx.begin(); }
+
+  template <typename T>
+  auto format(T const& t, auto& ctx) const {
+    auto out = std::format_to(ctx.out(), "{}{{", name_of(^T));
+
+    auto delim = [first=true, &out]() mutable {
+      if (!first) {
+        *out++ = ',';
+        *out++ = ' ';
+      }
+      first = false;
+    };
+
+    [: expand(bases_of(^T)) :] >> [&]<auto base>{
+        delim();
+        out = std::format_to(out, "{}", (typename [: type_of(base) :] const&)(t));
+    };
+
+    [: expand(nonstatic_data_members_of(^T)) :] >> [&]<auto mem>{
+      delim();
+      out = std::format_to(out, ".{}={}", name_of(mem), t.[:mem:]);
+    };
+
+    *out++ = '}';
+    return out;
+  }
+};
+
+struct B { int m0 = 0; };
+struct X : B { int m1 = 1; };
+struct Y : B { int m2 = 2; };
+class Z : public X, private Y { int m3 = 3; int m4 = 4; };
+
+template <> struct std::formatter<B> : universal_formatter { };
+template <> struct std::formatter<X> : universal_formatter { };
+template <> struct std::formatter<Y> : universal_formatter { };
+template <> struct std::formatter<Z> : universal_formatter { };
+
+int main() {
+    std::println("{}", Z());
```

Comment:
I think it would useful to test that this outputs what we expect, which we can do through certain annotations. Check out this test:
https://github.com/bloomberg/clang-p2996/blob/p2996/libcxx/test/std/experimental/reflection/p2996-ex-parsing-command-line-options-2.sh.cpp#L166

I think three steps ought to suffice:
1. Rename from `*.pass.cpp` to `*.sh.cpp`.
2. Add explicit `RUN` directives at the top of this file, akin to those found in [L19-20](https://github.com/bloomberg/clang-p2996/blob/121dc1100c41830d4403359bb20147be972d2026/libcxx/test/std/experimental/reflection/p2996-ex-parsing-command-line-options-2.sh.cpp#L19-L20) in the file linked above.
3. Add an explicit `RUN grep` directive above this `std::println` call to assert that the expected output is seen, akin to the one found in [L166](https://github.com/bloomberg/clang-p2996/blob/121dc1100c41830d4403359bb20147be972d2026/libcxx/test/std/experimental/reflection/p2996-ex-parsing-command-line-options-2.sh.cpp#L166) of the file linked above.

Thanks for putting this together!

Reply:
> I think it would useful to test that this outputs what we expect, which we can do through certain annotations. Check out this test: https://github.com/bloomberg/clang-p2996/blob/p2996/libcxx/test/std/experimental/reflection/p2996-ex-parsing-command-line-options-2.sh.cpp#L166
> 
> I think three steps ought to suffice:
> 
>     1. Rename from `*.pass.cpp` to `*.sh.cpp`.
> 
>     2. Add explicit `RUN` directives at the top of this file, akin to those found in [L19-20](https://github.com/bloomberg/clang-p2996/blob/121dc1100c41830d4403359bb20147be972d2026/libcxx/test/std/experimental/reflection/p2996-ex-parsing-command-line-options-2.sh.cpp#L19-L20) in the file linked above.
> 
>     3. Add an explicit `RUN grep` directive above this `std::println` call to assert that the expected output is seen, akin to the one found in [L166](https://github.com/bloomberg/clang-p2996/blob/121dc1100c41830d4403359bb20147be972d2026/libcxx/test/std/experimental/reflection/p2996-ex-parsing-command-line-options-2.sh.cpp#L166) of the file linked above.
> 
> 
> Thanks for putting this together!

I updated the test